src/main/python/deep_qa/models/reading_comprehension/gated_attention_reader.py

```diff
@@ -0,0 +1,271 @@
+            # concatenate qd_common_feature with document embeddings
+            # shape: (batch size, document_length, (2*seq2seq hidden size) + 2)
+            document_embedding = merge([document_embedding, qd_common_feature],
+                                       mode='concat')
+        # We encode the document embeddings with a final seq2seq encoder.
+        document_encoder = self._get_seq2seq_encoder(name="document_final")
+        # shape: (batch size, document_length, 2*seq2seq hidden size)
+        final_encoded_document = document_encoder(document_embedding)
+
+        if self.cloze_token is None:
+            # Get a final encoding of the question from a biGRU that does not return
+            # the sequence, and use it to calculate attention over the document.
+            final_question_encoder = self._get_encoder(name="question_final")
+            # shape: (batch size, 2*seq2seq hidden size)
+            final_encoded_question = final_question_encoder(question_embedding)
+        else:
+            # get a final encoding of the question by concatenating the forward
+            # and backward GRU at the index of the cloze token.
+            final_question_encoder = self._get_seq2seq_encoder(name="question_final")
+            # each are shape (batch size, question_length, seq2seq hidden size)
+            encoded_question_f, encoded_question_b = final_question_encoder(question_embedding)
+            # extract the gru outputs at the cloze token from the forward and
+            # backwards passes
+            index_selector = BiGRUIndexSelector(self.cloze_token_index)
+            final_encoded_question = index_selector([question_indices,
+                                                     encoded_question_f,
+                                                     encoded_question_b])
+
+        # take the softmax of the document_embedding after it has been passed
+        # through gated attention layers to get document probabilities
+        # shape: (batch size, document_length)
+        document_probabilities = Attention(name='question_document_softmax')([final_encoded_question,
+                                                                              final_encoded_document])
+        # We sum together the weights of words that match each option
+        # and use the multiword_option_mode to determine how to calculate
+        # the total probability of the option.
+        options_sum_layer = OptionAttentionSum(self.multiword_option_mode,
+                                               name="options_probability_sum")
+        # shape: (batch size, num_options)
+        options_probabilities = options_sum_layer([document_indices,
+                                                   document_probabilities,
+                                                   options_indices])
+
+        # We normalize the option_probabilities by dividing each
+        # element by L1 norm (sum) of the whole tensor.
```

Comment:
This confused me a bit - it's not the L1 norm of the whole tensor, it's per-batch. Given this is a common operation (batch normalisation) perhaps just change the wording here slightly?